### PR TITLE
Luna keyboard pet OLED timeout fix

### PR DESCRIPTION
**Pull request by joukewitteveen**
The old keyboard pet code would cause the screen to flicker after timeout because the animation would keep on running and each new frame would turn the screen on, after which the timeout logic would turn it off (provided the screen was idle for `> OLED_TIMEOUT`). The same flicker would be triggered by other triggers that turn the screen on, for instance toggling Caps Lock.

## Description

The on/off bug is fixed by using the animation idle timer (`anim_sleep`) also to track whether the screen is off or not. If the screen turns on (for whatever reason), we restart the timer. Also, we only render the keyboard pet when the screen is on, to prevent causing it to turn on.

## Implications and Alternatives

The keyboard pet OLED timeout logic duplicates the timeout logic in the OLED driver. There are two more thorough solutions I can think of that make reimplementing the logic unnecessary.

1. If the wpm counter drops to 0, we could render a single frame image of a sleeping pet. As this image doesn't change, the screen doesn't update and the timeout isn't reset each time the pet is rendered.
2. Support for `OLED_IDLE_BLOCKS_MASK` is easy to implement. This macro would specify a `OLED_BLOCK_TYPE` value and blocks in the mask would not cause the screen to be turned on (and the timeout timer to be reset).

For (2), the `oled_on()` line in `drivers/oled/ssd1306_sh1106.c:oled_render` would change to
```c
    if ((OLED_BLOCK_TYPE)1 << update_start) & ~((OLED_BLOCK_TYPE) OLED_IDLE_BLOCKS_MASK) {
        oled_on();
    }
```
This second option would offer a generic solution for using the screen timeout when a long-running animation is being shown. The downside is that figuring out the correct blocks mask is not very ergonomic. It is also a bit of a misuse of the block-based rendering logic.

As a remnant of looking into this, I had the addition of an assertion in the OLED driver implementation laying around. I added it to this PR. The assertion is violated in [code](https://github.com/qmk/qmk_firmware/blob/0.16.9/keyboards/rate/pistachio_pro/config.h) by @7-rate, but indeed: that code is broken since support for fonts with lines of any height other than 8 pixels is not implemented.

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Screen on/off bug with the Luna keyboard pet by @HellSingCoder.

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: drivers/oled/ssd1306_sh1106.c [hunk 1]
```diff
@@ -367,6 +367,8 @@ void oled_advance_page(bool clearPageRemainder) {
     uint16_t index     = oled_cursor - &oled_buffer[0];
     uint8_t  remaining = oled_rotation_width - (index % oled_rotation_width);
 
+    _Static_assert(OLED_FONT_HEIGHT == 8, "OLED buffer implementation assumes a line height of 8 pixels");
```
Comment (drashna) [inline drivers/oled/ssd1306_sh1106.c:370]: Not really a fan of this, as you *could* use multiples of 8, as well (considering the driver supports 128x64 screens, already)...

Also, this should be separated out into it's own PR, not tossed in with keyboard changes.

**Review by drashna — CHANGES_REQUESTED** [review]

File: drivers/oled/ssd1306_sh1106.c (hunk 1 above)
Comment (joukewitteveen) [reply drivers/oled/ssd1306_sh1106.c:370]: Multiples of 8 are not currently supported, but I agree that adding support for it would be easier than supporting arbitrary line heights.

I also agree that this change doesn't belong here, but I want to wait a bit with separating it from this PR since this PR could turn into a discussion about the OLED driver. In case the proposed `OLED_IDLE_BLOCKS_MASK` logic is desirable, this entire PR needs to be rewritten and will be more about the OLED driver than about the keyboard changes.

Comment (joukewitteveen) [reply drivers/oled/ssd1306_sh1106.c]: I removed the OLED related changes from this PR since I found an approach that does not require them. However, I do think it would be good to document the fact that `OLED_FONT_HEIGHT` is not really supported. Currently, it only says "untested", but really it should say "values other than 8 are not supported". I have no reason to work on the OLED driver anymore, so I'll leave that to those involved (@7-rate?).

**Comment by joukewitteveen** [comment]
@alexmalott: In your recent #17835 you must have encountered the screen flickering bug. What do you think of the alternative fix proposed here?

**Review by fauxpark — APPROVED** [review]

**Review by drashna — APPROVED** [review]
